pkg/registry/apis/secret/encryption/cipher/provider/cipher_aesgcm.go

```diff
@@ -0,0 +1,119 @@
+package provider
+
+import (
+	"context"
+	"crypto/aes"
+	cpr "crypto/cipher"
+	"crypto/rand"
+	"io"
+
+	"github.com/grafana/grafana/pkg/registry/apis/secret/encryption/cipher"
+)
+
+var (
+	_ cipher.Encrypter = (*aesGcmCipher)(nil)
+	_ cipher.Decrypter = (*aesGcmCipher)(nil)
+)
+
+type aesGcmCipher struct {
+	// randReader is used to generate random bytes for the nonce.
+	// This allows us to change out the entropy source for testing.
+	randReader io.Reader
+}
+
+func newAesGcmCipher() aesGcmCipher {
+	return aesGcmCipher{
+		randReader: rand.Reader,
+	}
+}
+
+func (c aesGcmCipher) Encrypt(_ context.Context, payload []byte, secret string) ([]byte, error) {
+	salt, err := c.generateSalt()
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := cipher.KeyToBytes([]byte(secret), salt)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	gcm, err := cpr.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	nonce, err := c.generateNonce(gcm)
+	if err != nil {
+		return nil, err
+	}
+
+	ciphertext := gcm.Seal(nil, nonce, payload, nil)
+
+	// Salt         Nonce          Encrypted
+	// │              │             Payload
+	// │              V               │
+	// │  ┌───────────────────────┐   │
+	// └►│SSSSSSSNNNNNNNEEEEEEEEE│◄─┘
+	//    └───────────────────────┘
+	prefix := append(salt, nonce...)
+	ciphertext = append(prefix, ciphertext...)
+
+	return ciphertext, nil
+}
+
+func (c aesGcmCipher) Decrypt(_ context.Context, payload []byte, secret string) ([]byte, error) {
```

Comment:
A nit I guess
```suggestion
func (c aesGcmCipher) Decrypt(_ context.Context, payload []byte, ciphertext string) ([]byte, error) {
```

Reply:
The payload is the ciphertext; the secret is the pre-shared agreed-upon key